src/bin/mv.rs

```diff
@@ -1,16 +1,404 @@
+use std::io::{self, BufRead, Read, Write, Stderr, StdinLock, StdoutLock};
+use std::os::unix::fs::MetadataExt;
+use std::path::{Path, PathBuf};
+use std::process::exit;
+use walkdir::WalkDir;
+
+const MAN_PAGE: &'static str = r#"NAME
+    mv - move (rename) files
+
+SYNOPSIS
+    mv [-i | --interactive] [-n | --no-clobber] [-v | --verbose] [h | --help] SOURCES.. DESTINATION
+
+DESCRIPTION
+    If a source is on the same device as it's respective destination, it will be renamed. If it is on a different device, it will be copied.
+
+    If the target is a directory, the source will be moved into that directory.
+
+OPTIONS
+    -h
+    --help
+        display this help and exit
+
+    -i
+    --interactive
+        prompt before overwriting existing files
+
+    -n
+    --no-clobber
+        do not overwrite existing files
+
+    -v
+    --verbose
+        print the file changes that have been successfully performed
+
+AUTHOR
+    Written by Michael Murphy.
+"#;
+
+/// Contains the sources, target and flags that were given as input arguments.
+struct Arguments {
+    sources: Vec<PathBuf>,
+    target:  PathBuf,
+    flags:   Flags
+}
+
+/// Stores the state of each flag.
+struct Flags {
+    interactive: bool,
+    noclobber:   bool,
+    verbose:     bool,
+}
 
 fn main() {
-    let mut stderr = stderr();
-    let ref src = env::args().nth(1).fail("no source argument.", &mut stderr);
-    let ref dst = env::args().nth(2).fail("no destination argument.", &mut stderr);
+    let stderr = &mut io::stderr();
+    let stdout = io::stdout();
+    let stdout = &mut stdout.lock();
+    let arguments = env::args().skip(1).collect::<Vec<String>>();
+    mv(check_arguments(&arguments, stdout, stderr), stdout, stderr);
+}
+
+/// Take a list of arguments and attempt to move each source argument to their respective destination.
+fn mv(arguments: Arguments, stdout: &mut StdoutLock, stderr: &mut Stderr) {
+    for source in arguments.sources {
+        // Metadata from the source and target are required to determine both are on the same device.
+        let target_metadata = get_target_metadata(&arguments.target, &source, stderr);
+        let target = get_target_path(&arguments.target, &target_metadata, &source, stderr);
+        let source_metadata = match get_source_metadata(&source, stderr) {
+            Some(metadata) => metadata,
+            None => continue // We will skip this source because there was an error.
+        };
+
+        // Move the source file or directory to the target path.
+        // If the source and target are on the same device, rename the source.
+        // If they are on different devices, copy the file or directory.
+        if source_metadata.dev() == target_metadata.dev() {
+            match fs::rename(&source, &target) {
+                Ok(_) => if arguments.flags.verbose { verbose_print(&source, &target, stdout); },
+                Err(message) => {
+                    let _ = stderr.write(b"cannot rename '");
```

Comment:
That is by purpose. It shouldn't throw an error message and then abort just because the abort message couldn't be send, since that could make it stuck in an infinite loop.

That is not possible in this (and the other) cases, though.